doc/design/block.md

```diff
@@ -0,0 +1,173 @@
+
+Blocks do not only group source code, but also narrow the lexical scope of variables so that they do not conflict with variables having the same name used elsewhere in a program.
+
+In Paddle, we need a similar concept called Block to support following scenes:
+
+- define a PaddlePaddle program by writing blocks of codes, which includes the definitions of variables and operators.
+  - `RNNOp`, `SwitchOp`, `WhileOp` and `IfElseOp` needs Block to help to define sub-block. 
+- help to execute multiple operators, blocks should group operators and runs like a single operator.
+
+## How to use Block
+In `RNNOp`, `SwitchOp`, `WhileOp` and `IfElseOp`, a sub-block should be used to help to define a sub-block.
+
+Let's start from how a RNNOp is described using Block:
+
+```python
+v = some_op()
+m_boot = some_op()
+
+W = pd.Variable(shape=[20, 20])
+U = pd.Varable(shape=[20, 20])
+
+rnn = RNNOp()
+with rnn.stepnet() as net:
+  x = net.add_input(v)
+  h = net.add_memory(init=m_boot)
+  
+  fc_out = pd.matmul(W, x)
+  hidden_out = pd.matmul(U, h)
+  sum = pd.add_two(fc_out, hidden_out)
+  act = pd.sigmoid(sum)
+  # declare outputs
+  net.add_output(act, hidden_out)
+
+acts, hs = rnn()
+```
+
+This python program will be transformed into Protobuf messages which describe the model, a C++ Block class will create the corresponding Variables and Operators, and execute all the operators.
+
+## Block Implementation
+
+During the Protobuf message generation, the Block should store VarDesc (the Protobuf message which describes Variable) and OpDesc (the Protobuf message which describes Operator).
+
+VarDesc in a block should have its name scope to avoid local variables affect father block's name scope. 
+Child block's name scopes should inherit the father's so that OpDesc in child block can reference a VarDesc that stored in father block. For example
+
+```python
+a = pd.Varaible(shape=[20, 20])
+b = pd.fc(a, params=["fc.w", "fc.b"])
+
+rnn = pd.create_RNNOp()
+with rnn.stepnet() as net:
+    x = net.add_input(a)
+    # reuse fc's parameter
+    fc_without_b = pd.get_variable("fc.w")
+    net.add_output(fc_without_b)
+
+out = rnn()
+```
+the method `pd.get_variable` can help retrieve a Variable by a name, a Variable may store in a father block, but might be retrieved in a child block, so block should have a variable scope that supports inheritance.
+
+We can implement this idea in the following approach
+
+```c++
+class Block {
+public:
+  Block(Block* father) : father_block_{father} {}
+  // add a new VarDesc, return the pointer to enable other functions.
+  // NOTE Will check whether some variable called the same name.
+  VarDesc* AddVarDesc();
+  OpDesc* AddOpDesc();
+
+  // name: variable's name
+  // recursive: whether to find the variable in father's scope recursively.
+  VarDesc* FindVarDesc(const string& name, bool recursive=true);
+
+private:
+  Block* father_block_;
+  // descriptions
+  map<VarDesc> var_descs_;
```

Comment:
var_descs_ => vars_
op_descs => ops_

Reply:
this example is a code snippet of defining the Protobuf in Block, `var_descs_` and `op_descs_` store description.

`ops_` member will be used to store `OperatorBase*`, and `vars_` member will be used to store `Variable*`.